internal/upload/oci/upload.go

```diff
@@ -0,0 +1,234 @@
+	}
+
+	ctx := context.Background()
+	resp, err := uploadManager.UploadFile(ctx, req)
+
+	if err != nil {
+		if resp.IsResumable() {
+			resp, err = uploadManager.ResumeUploadFile(ctx, *resp.MultipartUploadResponse.UploadID)
+			if err != nil {
+				return "", err
+			}
+		} else {
+			return "", fmt.Errorf("failed to upload the file to object %s:  %w", objectName, err)
+		}
+
+	}
+	// clean up the object even if we fail
+	defer c.deleteObjectFromBucket(objectName, bucketName, namespace)
+
+	imageID, err := c.create(objectName, bucketName, namespace, compartmentID, imageName)
+	if err != nil {
+		return "", fmt.Errorf("failed to create a custom image using object '%s' bucket '%s' in namespace '%s': %w",
+			objectName,
+			bucketName,
+			namespace,
+			err)
+	}
+	err = c.GrantImageAccess(imageID, userID, compartmentID)
+	if err != nil {
+		return "", fmt.Errorf("failed to grant image ID '%s' access to user %s in compartment %s: ",
+			imageID,
+			userID,
+			compartmentID)
+	}
+
+	return imageID, nil
+}
+
+// Create creates an image from the storageObjectName stored in the bucketName.
+// The result is an image ID or an error if the operation failed.
+func (c Client) create(objectName, bucketName, namespace, compartmentID, imageName string) (string, error) {
+	request := core.CreateImageRequest{
+		CreateImageDetails: core.CreateImageDetails{
+			DisplayName:   common.String(imageName),
+			CompartmentId: common.String(compartmentID),
+			FreeformTags: map[string]string{
+				"Uploaded-By": "osbuild-composer",
+				"Name":        imageName,
+			},
+			ImageSourceDetails: core.ImageSourceViaObjectStorageTupleDetails{
+				BucketName:      common.String(bucketName),
+				NamespaceName:   common.String(namespace),
+				ObjectName:      common.String(objectName),
+				SourceImageType: core.ImageSourceDetailsSourceImageTypeQcow2,
+			},
+		},
+	}
+
+	createImageResponse, err := c.computeClient.CreateImage(context.Background(), request)
+	if err != nil {
+		return "", fmt.Errorf("failed to create an image from storage object: %w", err)
+	}
+
+	log.Printf("waiting for the work request to complete, this may take a while. Work request ID: %s\n", *createImageResponse.OpcWorkRequestId)
+	for {
+		r, err := c.workRequestsClient.GetWorkRequest(context.Background(), workrequests.GetWorkRequestRequest{
+			WorkRequestId: createImageResponse.OpcWorkRequestId,
+			OpcRequestId:  createImageResponse.OpcRequestId,
+		})
+		if err != nil {
+			return "", fmt.Errorf("failed to fetch the work request for creating the image: %w", err)
+		}
+		switch r.Status {
+		case workrequests.WorkRequestStatusSucceeded:
+			return *createImageResponse.Id, nil
+		case workrequests.WorkRequestStatusCanceled, workrequests.WorkRequestStatusFailed:
+			return "", fmt.Errorf("the work request for creating an image is in status %s", r.Status)
+		}
+		log.Printf("work request - status: %s, percent completed: %f\n", r.Status, *r.PercentComplete)
+		time.Sleep(10 * time.Second)
```

Comment:
Perhaps it would be better to reduce the interval to 1 second (or less even) and remove the progress printing so as to not flood the log.

Reply:
ack

Reply:
any thoughts on having debug or other log levels? This is a valuable info to pass.
we can print the state only at fixed intervals, every 10/20 seconds and still loop each sec

Reply:
anyhow, meanwhile I reduced the interval to 1 sec